syncweek-react-aad/react-aad
# Add ability to get a valid/refreshed token for API request

## Purpose
Currently, the component allows you to use authentication successfully. However, it's not possible to use the underlying MSAL to get a fresh token to make a web request with. This review solves that limitation and refactors the code substantially to make future improvements possible and bring us in line with the MSAL Angular wrapper.

Changes:
- I've merged the Redirect and Popup authentication provider types into the same file. There wasn't enough logic different to justify separate classes, and it makes changing the login type later difficult. Things that aren't the MSAL configurations (`AuthenticationParameters`, `LoginType`, etc) should be able to be changed at any point.
- The new single `MsalAuthProvider` class extends the `UserAgentApplication` class from MSAL. This is how the MSAL Angular wrapper works because there are protected methods and fields in UserAgentApplication that you can't access otherwise.
- Add getter/setters for the `AuthenticationParameters` and `LoginType`.
- The `AzureAD` component is no longer responsible for dispatching redux actions. This should be the responsibility of the authentication provider itself so other components can reuse the same data. This also decouples the component from the provider.
- Added samples for fetching a valid token for API requests.
- Added additional Redux action types for login success/failure, token refresh success/error, and the AAD store initialization.

## Does this introduce a breaking change?
```
[ ] Yes
[x] No
```
There are no breaking changes, but there is stuff that no longer makes sense to have such as the auth factory, but it's left intact for backward compatibility.

## Pull Request Type
What kind of change does this Pull Request introduce?


```
[ ] Bugfix
[x] Feature
[x] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
-> Load the sample project
-> Authenticate with either of the samples
-> After successful login, the new "Get Token" button should appear. Click the button to get a valid token
```

## Other Information
I will send a separate PR for documentation changes. But I realized the documentation needs to be largely rewritten, and I'd prefer to open an issue to track that.
